src/custodian/vasp/jobs.py

```diff
@@ -704,19 +710,53 @@ def constrained_opt_run(
                 file.write(f"{key} {energies[key]}\n")
 
     def terminate(self, directory="./") -> None:
-        """Kill all VASP processes associated with the current job."""
+        """Kill all VASP processes associated with the current job.
+
+        On POSIX, sends SIGTERM/SIGKILL to entire process group to ensure all
+        MPI workers are terminated. On Windows, falls back to process.terminate().
+        """
+        pid = self._vasp_process.pid
+
         if self._vasp_process.poll() is not None:
-            logger.warning("The process was already done!")
+            logger.warning(f"Process {pid} already terminated")
             return
 
-        try:
-            logger.info(f"Killing PID {self._vasp_process.pid}")
+        # POSIX: kill entire process group; Windows: just terminate parent
+        use_pgid = hasattr(os, "killpg")
+        if use_pgid:
+            try:
+                pgid = os.getpgid(pid)
+            except ProcessLookupError:
+                logger.warning(f"Process {pid} not found (already dead)")
+                return
+            logger.info(f"Sending SIGTERM to process group {pgid}")
+            try:
+                os.killpg(pgid, signal.SIGTERM)
+            except ProcessLookupError:
+                logger.warning(f"Process group {pgid} not found (already dead)")
+                return
```

Comment:
Does this work as expected when VASP is run across multiple nodes? It seems that a similar approach in https://github.com/materialsproject/custodian/pull/231 did not work as expected. I do not know the details offhand.